force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,191 @@
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2021
+* @description Donation History Service class in NPSP.
+*/
+public with sharing class DonationHistoryService {
+
+    static final String HARD_CREDIT = 'Hard Credit';
+    static final String SOFT_CREDIT = 'Soft Credit';
+    static final String PARTIAL_SOFT_CREDIT = 'Partial Soft Credit';
+    static final String NPSP_CREDIT_ROLES= 'NPSP_OppsSoftCreditRoles';
+    
+    @TestVisible
+    private UTIL_Permissions permissions {
+        get {
+            if(permissions == null) {
+                permissions = new UTIL_Permissions();
+            }
+            return permissions;
+        }
+        set;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return Donation History Access verification.
+    */
+    private Boolean hasAccessToDonationHistory(){
+        return permissions.canRead(Opportunity.SObjectType, new Set<SObjectField>{
+            Opportunity.Amount,
+            Opportunity.CloseDate,
+            Opportunity.Name
+        }) && permissions.canRead(Contact.SObjectType, new Set<SObjectField>{
+            Contact.Email,
+            Contact.Name
+        }) && permissions.canRead(Partial_Soft_Credit__c.SObjectType, new Set<SObjectField>{
+            Partial_Soft_Credit__c.Amount__c,
+            Partial_Soft_Credit__c.Contact_Role_ID__c
+        });
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description method to return Donation History list.
+    */
+    public List<DonationHistoryController.DonationHistory> getDonationHistory(Id contactId){
+        if(hasAccessToDonationHistory()){
+            List<OpportunityContactRole> credits = new DonationHistorySelector().getOpportunityContactRoleList(contactId);
+            List<DonationHistoryController.DonationHistory> result = buildDonationHistoryFromOCRs(credits);
+            return result;
+        } else {
+            throw new UTIL_Permissions.InsufficientPermissionException(Label.commonInsufficientPermissions);
+        }
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify each condition to list Donation.
+    */
```

Comment:
Generally we don't write ApexDoc comments for private methods. The Function name, Signature, and Return type here are pretty descriptive on their own.

Reply:
done